### DISCUSSION TITLE
Kernel error in Crow app

### DISCUSSION THREAD

**Discussion by Giovar6** [idaholab/moose#16408] [Q&A Modules: Phase field] [answered]
Hello Moose community,

I would like to introduce myself first. I am Giovanni and come form Italy. Currently I am working on metal sintering, and in particular on solid state metal sintering simulation.

I've found that Sudipta Biswas has already done something similar on moose with the Crow application (https://github.com/SudiptaBiswas/Crow). I see that a noticeable work has been done and I would like to start from these applications to develop my project. I have copied the following input file in my own application and also the other objects necessary to run correctly this application.
Here is the input file:

```
[GlobalParams]
  var_name_base = gr
  op_num = 2.0
  int_width = 2.0
  #en_ratio = 1
[]

[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 80
  ny = 40
  nz = 0
  xmin = 0.0
  xmax = 40.0
  ymin = 0.0
  ymax = 20.0
  zmax = 0
  elem_type = QUAD4
[]

[Variables]
  [./c]
    #scaling = 10
  [../]
  [./w]
  [../]
  [./PolycrystalVariables]
  [../]
[]

[ICs]
  [./ic_gr1]
    int_width = 2.0
    x1 = 25.0
    y1 = 10.0
    radius = 7.4
    outvalue = 0.0
    variable = gr1
    invalue = 1.0
    type = SmoothCircleIC
  [../]
  [./multip]
    x_positions = '10.0 25.0'
    int_width = 2.0
    z_positions = '0 0'
    y_positions = '10.0 10.0 '
    radii = '7.4 7.4'
    3D_spheres = false
    outvalue = 0.001
    variable = c
    invalue = 0.999
    type = SpecifiedSmoothCircleIC
    block = 0
  [../]
  [./ic_gr0]
    int_width = 2.0
    x1 = 10.0
    y1 = 10.0
    radius = 7.4
    outvalue = 0.0
    variable = gr0
    invalue = 1.0
    type = SmoothCircleIC
  [../]
[]

[AuxVariables]
  [./bnds]
  [../]
  [./total_en]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./unique_grains]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./var_indices]
    order = CONSTANT
    family = MONOMIAL
  [../]
  [./centroids]
    order = CONSTANT
    family = MONOMIAL
  [../]
[]

[Kernels]
  [./cres]
    type = SplitCHParsed
    variable = c
    kappa_name = kappa_c
    w = w
    f_name = F
    args = 'gr0 gr1'
  [../]
  [./wres]
    type = SplitCHWRes
    variable = w
    mob_name = D
  [../]
  [./time]
    type = CoupledTimeDerivative
    variable = w
    v = c
  [../]
  [./PolycrystalSinteringKernel]
    c = c
    consider_rigidbodymotion = false
    anisotropic = false
    grain_force = grain_force
    grain_tracker_object = grain_center
    grain_volumes = grain_volumes
    translation_constant = 10.0
    rotation_constant = 1.0
  [../]
[]

[AuxKernels]
  [./bnds]
    type = BndsCalcAux
    variable = bnds
    v = 'gr0 gr1'
  [../]
  [./Total_en]
    type = TotalFreeEnergy
    variable = total_en
    kappa_names = 'kappa_c kappa_op kappa_op'
    interfacial_vars = 'c  gr0 gr1'
  [../]
  [./unique_grains]
    type = FeatureFloodCountAux
    variable = unique_grains
    flood_counter = grain_center
    field_display = UNIQUE_REGION
    execute_on = timestep_begin
  [../]
  [./var_indices]
    type = FeatureFloodCountAux
    variable = var_indices
    flood_counter = grain_center
    field_display = VARIABLE_COLORING
    execute_on = timestep_begin
  [../]
  [./centroids]
    type = FeatureFloodCountAux
    variable = centroids
    execute_on = timestep_begin
    field_display = CENTROID
    flood_counter = grain_center
  [../]
[]

[BCs]
   [./flux]
     type = CahnHilliardFluxBC
     variable = w
     boundary = 'top bottom left right'
     flux = '0 0 0'
     mob_name = D
     args = 'c'
   [../]
[]

[Functions]
  [./load]
    type = ConstantFunction
    value = 0.01
  [../]
[]

[Materials]
  [./free_energy]
    type = SinteringFreeEnergy
    block = 0
    c = c
    v = 'gr0 gr1'
    f_name = S
    derivative_order = 2
    outputs = console
  [../]
  [./CH_mat]
    type = PFDiffusionGrowth
    block = 0
    rho = c
    v = 'gr0 gr1'
    outputs = console
  [../]
  [./constant_mat]
    type = GenericConstantMaterial
    block = 0
    prop_names = '  A    B   L   kappa_op kappa_c'
    prop_values = '16.0 1.0 1.0  0.5      1.0    '
  [../]
[]

[VectorPostprocessors]
  [./grain_volumes]
    type = FeatureVolumeVectorPostprocessor
    flood_counter = grain_center
    execute_on = 'initial timestep_begin'
  [../]
[]

[UserObjects]
  [./grain_center]
    type = GrainTracker
    outputs = none
    compute_var_to_feature_map = true
    execute_on = 'initial timestep_begin'
  [../]
[]

[Postprocessors]
  [./elem_c]
    type = ElementIntegralVariablePostprocessor
    variable = c
  [../]
  [./elem_bnds]
    type = ElementIntegralVariablePostprocessor
    variable = bnds
  [../]
  [./total_energy]
    type = ElementIntegralVariablePostprocessor
    variable = total_en
  [../]
  [./free_en]
    type = ElementIntegralMaterialProperty
    mat_prop = F
  [../]
  [./dofs]
    type = NumDOFs
    system = 'NL'
  [../]
  [./tstep]
    type = TimestepSize
  [../]
   [./run_time]
    type = RunTime
    time_type = active
   [../]
  [./int_area]
    type = InterfaceAreaPostprocessor
    variable = c
  [../]
  [./grain_size_gr0]
    type = ElementIntegralVariablePostprocessor
    variable = gr0
  [../]
  [./grain_size_gr1]
    type = ElementIntegralVariablePostprocessor
    variable = gr1
  [../]
  [./gb_area]
    type = GrainBoundaryArea
  [../]
[]

[Preconditioning]
  [./SMP]
    type = SMP
    coupled_groups = 'c,w c,gr0,gr1 '
  [../]
[]

[Executioner]
   Preconditioned JFNK (default)
  type = Transient
  scheme = BDF2
  solve_type = NEWTON
  petsc_options_iname = '-pc_type -ksp_grmres_restart -sub_ksp_type -sub_pc_type -pc_asm_overlap'
  petsc_options_value = 'asm         31   preonly   lu      1'
  l_max_its = 20
  nl_max_its = 20
  nl_abs_tol = 1e-10
  nl_rel_tol = 1e-08
  l_tol = 1e-04
  end_time = 20
  #dt = 0.01
  [./TimeStepper]
    type = IterationAdaptiveDT
    dt = 0.01
    growth_factor = 1.2
  [../]
[]

[Adaptivity]
  marker = bound_adapt
  max_h_level = 2
  [./Indicators]
    [./error]
      type = GradientJumpIndicator
      variable = bnds
    [../]
  [../]
  [./Markers]
    [./bound_adapt]
      type = ValueRangeMarker
      lower_bound = 0.01
      upper_bound = 0.99
      variable = bnds
    [../]
  [../]
[]

[Outputs]
  print_linear_residuals = true
  csv = true
  gnuplot = true
  perf_graph = true
  [./console]
    type = Console
    PerfGraphOutput = true
  [../]
  [./exodus]
    type = Exodus
    elemental_as_nodal = true
  [../]
[]

[Debug]
  show_var_residual_norms = true
[]
```

**but when I run this input file with peacock i get this error:**

WARNING
 Error producing mesh: '/home/giovanni/projects/lucky/lucky-opt -i /home/giovanni/projects/lucky/problems/peacock_run_mesh_tmp.i --mesh-only /home/giovanni/projects/lucky/problems/peacock_run_mesh_tmp_Sintering.e --allow-test-objects' exited with non zero status 1.

Please make sure your application is built and able to execute the given arguments.
Working dir: /home/giovanni/projects/lucky/problems
Output: 

```
*** ERROR ***
/home/giovanni/projects/lucky/problems/peacock_run_mesh_tmp.i:108: missing required parameter 'Kernels/PolycrystalSinteringKernel/type'
	Doc String: "A string representing the Moose Object that will be built by this Action"


Stack frames: 16
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&&)
3: InputParameters::checkParams(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
4: ActionFactory::create(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, InputParameters&)
5: Parser::walkRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, hit::Node*)
6: Parser::walk(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, hit::Node*)
7: hit::Node::walk(hit::Walker*, hit::NodeType)
8: hit::Node::walk(hit::Walker*, hit::NodeType)
9: hit::Node::walk(hit::Walker*, hit::NodeType)
10: Parser::parse(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
11: MooseApp::setupOptions()
12: MooseApp::run()
13: /home/giovanni/projects/lucky/lucky-opt(+0x51d1) [0x55ccce13e1d1]
14: __libc_start_main
15: /home/giovanni/projects/lucky/lucky-opt(+0x54af) [0x55ccce13e4af]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor

Exodus filenames: ['peacock_run_exe_tmp_Sintering_exodus.e']
Postprocessor filenames: ['peacock_run_exe_tmp_Sintering_out.csv']
VectorPostprocessor filenames: ['peacock_run_exe_tmp_Sintering_out_grain_volumes_*.csv']
Running command: /home/giovanni/projects/lucky/lucky-opt Outputs/csv=true -i peacock_run_exe_tmp_Sintering.i
Working directory: /home/giovanni/projects/lucky/problems
Finished: Exit code: 1
```

Now, according to the error description I should specify the type of PolycrystalSinteringKernel, but in the original input file it was not specified and the simulation runs correctly. I would like to precise that the application is correctly compiled and that I have performed a successful test.
Can anyone please explain where I am wrong, if possible?
Probably is a stupid question, but I am at my beginning with moose and not everything is clear.
Thank you in advance for your answer.

**Comment by permcody** [comment, marked as answer]
I would suggest you start by just cloning that repository directly and making sure that you can build and run from it. It appears that the "devel" branch is reasonably current, but if you pulled from "master" or one of the other branches the source is going to be very out of date. Did you copy from "devel"?

To get a little closer to answering your question though, the problem is likely to do with registration of that sintering "Action". This is custom syntax that creates multiple objects with one block but since it's not registered in your application properly, MOOSE thinks you are just adding in some "single" kernel and made a syntax error (e.g. "type" is required in most basic input blocks and usually corresponds to the same named C++ source file). See if you get get her app working directly, then you can start moving things over. The other alternative would be to work with her directly and add to her application. She may be willing to collaborate with you directly in that application.

**Reply by Giovar6** [reply]
Hello and thank you for your fast reply.

To answer your question, I have directly forked the devel branch of the crow application and I was able to build perfectly. I was also able to run the code properly, from the devel branch of the Crow application.
I believe you identified the problem. How can I make MOOSE understand that this is not a "single" kernel? How can I properly define this Action?
Thank you for your answer!

**Reply by permcody** [reply]
Registering new syntax is one of the few things that does not occur right in the new object source file. It occurs in the main application: https://github.com/SudiptaBiswas/Crow/blob/7402c20b48166c7f75392ba9b2101e8043c01fdc/src/base/CrowApp.C#L32-L33

While you are looking at the "App" take note of the other things that occur here (i.e. the registration of the required modules and other application dependencies). If you follow this pattern in your application you should be up and running.

**Reply by Giovar6** [reply]
Thank you for your answer. 
I've tried to follow your suggestions, but still was not able to run. I want to attempt something else. If not able to run I will write again.

**Comment by dschwen** [comment]
Tagging @SudiptaBiswas

**Reply by Giovar6** [reply]
Thank you for your answer and for tagging Sudipta Biswas

**Reply by SudiptaBiswas** [reply]
Something amiss with my GitHub settings. I did not get a notification even after being tagged. @Giovar6 were you able to run this with peacock?

**Reply by Giovar6** [reply]
Hi @SudiptaBiswas. Yes, if I run the input file from the forked devel branch of your project I am able to run with peacock. If I copy the input file and the objects required by it in my project I am not able to run. I believe the suggestions of @permcody should help me to solve the problems, but still was not able to run.
Thank you for your interest

**Reply by SudiptaBiswas** [reply]
By your project do you mean, you have another MOOSE based application? If so, I would recommend coupling the applications rather than copying the files over. I could help you with that if needed.

**Reply by Giovar6** [reply]
Hi, yes I mean that I have copied the input file and other objects in my application. Thank you for your support. I would be very grateful for your support to couple the applications.

**Comment by Giovar6** [comment]
Hi all, And thank you for the support. Today I've tried again to follow the suggestions of @permcody and finally I was able to run the input file of the Crow app within my application. To solve the problem I've copied and edited to fit my application. the CrowApp.C and CrowApp.h files.
